File: v1/src/main/java/com/api/v1/domain/entities/Customer.java
package com.api.v1.domain.entities;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import com.api.v1.dtos.requests.CustomerRegistrationRequestDto;
import lombok.Getter;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

@Document(collection = "v1_customers")
@Getter
public class Customer {

    @Field
    private String createdAt;

    @Field
    private String firstName;

    @Field
    private String middleName;

    @Field
    private String lastName;

    @Field
    private String ssn;

    @Field
    private LocalDate birthDate;

    @Field
    private String email;

    @Field
    private String address;

    @Field
    private String phoneNumber;

    @Field
    private String gender;

    @Field
    private String updatedAt;

    public Customer(
        String firstName, 
        String middleName, 
        String lastName, 
        String ssn, 
        LocalDate birthDate, 
        String email,
        String address, 
        String phoneNumber, 
        String gender
    ) {
        this.createdAt = ZonedDateTime.now().toString();
        this.firstName = firstName;
        this.middleName = middleName;
        this.lastName = lastName;
        this.ssn = ssn;
        this.birthDate = birthDate;
        this.email = email;
        this.address = address;
        this.phoneNumber = phoneNumber;
        this.gender = gender;
    }

    public void update(CustomerRegistrationRequestDto request) {
        this.firstName = request.firstName();
        this.middleName = request.middleName();
        this.lastName = request.lastName();
        this.ssn = request.ssn();
        this.birthDate = request.birthDate();
        this.email = request.email();
        this.address = request.address();
        this.phoneNumber = request.phoneNumber();
        this.gender = request.gender();
        this.updatedAt = ZonedDateTime.now().toString();
    }

    public String getFullName() {
        if (middleName.isEmpty()) return String.format("%s %s", firstName, lastName);
        return String.format("%s %s %s", firstName, middleName, lastName);
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/customer/domain/Customer.kt
package com.api.v2.customer.domain

import com.api.v2.customer.dtos.CustomerModificationRequestDto
import com.api.v2.customer.dtos.CustomerRegistrationRequestDto
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.*

@Table("customers")
class Customer {

    @Id
    var id: UUID? = null
    lateinit var firstName: String
    var middleName: String? = null
    lateinit var lastName: String
    lateinit var ssn: String
    lateinit var birthDate: LocalDate
    lateinit var email: String
    lateinit var gender: String
    lateinit var phoneNumber: String
    var createdAt: Instant = Instant.now()
    var creationZoneId: ZoneId = ZoneId.systemDefault()
    var archivedAt: Instant? = null
    var archivingZoneId: ZoneId? = null

    constructor() {}

    constructor(requestDto: CustomerRegistrationRequestDto) {
        this.firstName = requestDto.firstName
        this.middleName = requestDto.middleName
        this.lastName = requestDto.lastName
        this.ssn = requestDto.ssn
        this.birthDate = requestDto.birthDate
        this.email = requestDto.email
        this.gender = requestDto.gender
        this.phoneNumber = requestDto.phoneNumber
    }

    constructor(
        ssn: String,
        requestDto: CustomerModificationRequestDto,
        createdAt: Instant,
        creationZoneId: ZoneId
    ) {
        this.firstName = requestDto.firstName
        this.middleName = requestDto.middleName
        this.lastName = requestDto.lastName
        this.ssn = ssn
        this.birthDate = requestDto.birthDate
        this.email = requestDto.email
        this.gender = requestDto.gender
        this.phoneNumber = requestDto.phoneNumber
        this.createdAt = createdAt
        this.creationZoneId = creationZoneId
    }

    fun fullName(): String {
        if (middleName.isNullOrEmpty()) {
            return "$firstName $lastName"
        }
        return "$firstName $middleName $lastName"
    }

    fun archive() {
        archivedAt = Instant.now()
        archivingZoneId = ZoneId.systemDefault()
    }

}
